Streams: Test for pipeThrough undefined guard

ricea opened the pull request:
Verify that pipeThrough() throws when the readable or writable arguments
are undefined.

Also verify that other types are passed through to pipeTo() and the
return value.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

ricea committed:
Streams: Test for pipeThrough undefined guard

Verify that pipeThrough() throws when the readable or writable arguments
are undefined.

Also verify that other types are passed through to pipeTo() and the
return value.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/258495502)

# Firefox (nightly) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/258495503)

# Sauce (safari) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/258495516)

# Chrome (unstable) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/258495517)

# Sauce (MicrosoftEdge) #

ricea commented:
This is a test of the new behaviour added in https://github.com/whatwg/streams/pull/759, so it doesn't pass yet.

domenic commented on streams/piping/pipe-through.js:
```diff
@@ -136,12 +138,37 @@ test(() => {
+  for (const arg of invalidArguments) {
+    const rs = new ReadableStream();
+    assert_equals(arg, rs.pipeThrough({ writable: arg, readable: arg }),
+                  'pipeThrough() should not throw for ' + JSON.stringify(arg));
```
So, these should throw because it will delegate to `rs.pipeTo(arg)`, which will throw since `IsWritableStream(arg)` is false.

We can change these tests to work on a `rs` with an overridden (never-throwing) `pipeTo` if we want.

domenic replied:
Nevermind this comment, it was definitely wrong. This is a good test.

The other comments apply, but are not as critical. So, I will switch this to approved to reduce review lag.

domenic commented on streams/piping/pipe-through.js:
```diff
@@ -136,12 +138,37 @@ test(() => {
+  for (const pair of pairs) {
+    const rs = new ReadableStream();
+    assert_throws(new TypeError(), () => rs.pipeThrough(pair),
+                  'pipeThrough should throw for argument ' + JSON.stringify(pair) + ';');
```
JSON.stringify will just ignore undefined properties, so several of these test cases will be indistinguishable. Maybe add the index of the failure too?

ricea replied:
Done.

domenic commented on streams/piping/pipe-through.js:
```diff
@@ -136,12 +138,37 @@ test(() => {
+  }
+}, 'undefined readable or writable arguments should cause pipeThrough() to throw');
+
+test(() => {
```
I think separate tests for `{ writable: new WritableStream(), readable: arg }` and the counterpart would be more interesting.

ricea replied:
Done.

ricea committed:
Remove expectation that pipeThrough works with missing streams

readable-streams/pipe-through.js had a test that pipeThrough() still worked with
missing readable and writable. Since this behaviour is changing, remove the
expectation.

Also split out the tests for unusual values of `readable` and `writable`, and
include the index in the assert message for undefined values.

ricea commented:
I found I missed a test which expected the old behaviour, so I changed it. PTAL.

domenic commented on streams/readable-streams/pipe-through.js:
```diff
@@ -98,11 +98,9 @@ test(() => {
   };
 
```
I think this file should get merged with piping/pipe-through one day. Not today.

ricea replied:
Indeed. It confused me that there are two, and I don't have a clear idea of which tests should go where.

tyoshino commented on streams/piping/pipe-through.js:
```diff
@@ -30,6 +30,8 @@ function duckTypedPassThroughTransform() {
   };
 }
 
+const uninterestingReadableWritablePair = { writable: new WritableStream(), readable: new ReadableStream() };
```
though currently this is fine, i suggest that this kind of shared test data to be created for each test cases in principle. how about making this a factory function?

ricea replied:
Done.

ricea committed:
Make uninterestingReadableWritablePair a function

uninterestingReadableWritablePair being a variable that was re-used meant that a
single malfunctioning test could cause other tests to malfunction.

Change it to a function that always returns a new value.
